### PR TITLE
Compile zlib static library with -fPIC

### PR DESCRIPTION
**Pull request by wesm**
Same issue as conda-forge/snappy-feedstock#1. With these patches I'm able to link a shared library (apache/parquet-cpp) that depends on static libs for both snappy and zlib.

### CODE REVIEW

**Comment by wesm** [comment]
Looks like appveyor is hung. Can we merge this in? Thanks

**Comment by jakirkham** [comment]
Actually, it [ran](https://ci.appveyor.com/project/conda-forge/zlib-feedstock/build/1.0.7), but the status isn't updated. We have seen this a couple of times, but don't know the cause. Appears there are some failures under VC9. Though it appears to be caused by downloading issues not necessarily a real problem here. Restarting...

**Comment by wesm** [comment]
Is this good to go?

**Comment by msarahan** [comment]
We should discuss whether changes that affect only one platform should bump the build number of all platforms.  Some people think that the build number should be uniform on all platforms, so that tools that record environments are more cross-platform.  I think it's silly to give such meaning to the build number, but I don't really care what we do.

Either way, IMHO we can merge this now and discuss it later.  @jakirkham, thoughts?

**Comment by jakirkham** [comment]
So, I looked a bit more closely and I think `-fPIC` is already used by default for `zlib` based on their `./configure` file. This can be seen in our existing Travis CI build [log](https://travis-ci.org/conda-forge/zlib-feedstock#L422) and CircleCI build [log](https://circleci.com/gh/conda-forge/zlib-feedstock/9).

As for the `-O3` flag, this is used with `gcc` (and can be seen in the CircleCI log), but is intentionally removed when using `clang` (demonstrated by the Travis CI log). I don't know why it is being removed, but am a bit leary of adding it if they remove it. Are we sure that this is actually needed? Do we know why they remove it with `clang`?

**Comment by jakirkham** [comment]
In short, am not ready to merge this. Am questioning why it is needed (w.r.t `-fPIC`) or whether it is a good idea (w.r.t. `-O3`).

**Comment by wesm** [comment]
@jakirkham that is true _only_ of the code used to produce the shared library -- the object code that is linked into the static library is not being compiled with -fPIC

https://travis-ci.org/conda-forge/zlib-feedstock#L418

(it is the same in the CircleCI build, too)

I don't have an opinion about the optimization flag; I do know that setting `CFLAGS` overrides whatever optimization flag is set otherwise by configure.

**Comment by jakirkham** [comment]
> @jakirkham that is true only of the code used to produce the shared library -- the object code that is linked into the static library is not being compiled with -fPIC
> 
> https://travis-ci.org/conda-forge/zlib-feedstock#L418
> 
> (it is the same in the CircleCI build, too)

So I followed up on your ref and tried to inspect what was going on with the static library. I agree with you that we need to explicitly set `-fPIC` for the static library.

> I don't have an opinion about the optimization flag...

I am just a little worried about this. I have seen some libraries behave poorly when some optimization levels and it is unclear to me whether it is an issue with the library or the compiler. I do see others building with `clang` and `-O3` particularly over at Homebrew. Still I see code in `./configure` that seems to add this flag and then strip it out in some cases. It would be preferable if we can figure out why it is getting stripped out instead of just overriding it. So for now, my vote is for dropping the optimization flag. If that sounds reasonable to you, let's make that change.

<br>
TL;DR - Switching to `CFLAGS="-fPIC"` is acceptable. Let's please hold off on `-O3` until we bette understand what is going on there better.

**Comment by wesm** [comment]
Done. Merge when ready (on green build, I guess)

**Comment by jakirkham** [comment]
As the AppVeyor queue is a bit backed up, this change doesn't affect Windows, and all other relevant CIs are passing, I'm happy to merge this now.

**Comment by wesm** [comment]
👍
